File: core/turing_machine.py
from dataclasses import dataclass, field
from typing import Dict, Tuple, Set, Optional

Move = str  # 'L' | 'R' | 'N'
Transition = Tuple[str, str, Move]


@dataclass
class TuringMachine:
    states: Set[str]
    input_symbols: Set[str]
    tape_symbols: Set[str]
    blank: str
    transitions: Dict[Tuple[str, str], Transition]
    start_state: str
    accept_states: Set[str]
    reject_states: Set[str]
    tape: Dict[int, str] = field(default_factory=dict)
    head: int = 0
    current_state: Optional[str] = None
    halted: bool = False
    # 'ACCEPT' | 'REJECT' | 'NO_TRANSITION' | 'MAX_STEPS'
    result: Optional[str] = None

    def reset(self, input_string: str):
        self.tape = {}
        for i, ch in enumerate(input_string):
            self.tape[i] = ch
        self.head = 0
        self.current_state = self.start_state
        self.halted = False
        self.result = None

    def read(self) -> str:
        return self.tape.get(self.head, self.blank)

    def write(self, symbol: str):
        if symbol == self.blank:
            if self.head in self.tape:
                del self.tape[self.head]
        else:
            self.tape[self.head] = symbol

    def step(self) -> None:
        if self.halted:
            return
        if self.current_state in self.accept_states:
            self.halted, self.result = True, 'ACCEPT'
            return
        if self.current_state in self.reject_states:
            self.halted, self.result = True, 'REJECT'
            return

        sym = self.read()
        key = (self.current_state, sym)
        if key not in self.transitions:
            self.halted = True
            self.result = 'NO_TRANSITION'
            return
        new_state, write_sym, move = self.transitions[key]
        self.write(write_sym)
        if move == 'L':
            self.head -= 1
        elif move == 'R':
            self.head += 1
        elif move == 'N':
            pass
        else:
            raise ValueError(f"Movimento inválido: {move}")
        self.current_state = new_state

    def run(self, max_steps: int = 1000):
        steps = 0
        while not self.halted and steps < max_steps:
            self.step()
            steps += 1
        if not self.halted and steps >= max_steps:
            self.halted = True
            self.result = 'MAX_STEPS'
        return steps

    def window_cells(self, span: int = 25):
        left = self.head - span
        right = self.head + span
        return [(i, self.tape.get(i, self.blank), (i == self.head)) for i in range(left, right + 1)]

    def to_dict(self):
        """Converte para dicionário serializável"""
        return {
            'states': list(self.states),
            'input_symbols': list(self.input_symbols),
            'tape_symbols': list(self.tape_symbols),
            'blank': self.blank,
            'transitions': {f"{k[0]},{k[1]}": list(v) for k, v in self.transitions.items()},
            'start_state': self.start_state,
            'accept_states': list(self.accept_states),
            'reject_states': list(self.reject_states),
            'tape': {str(k): v for k, v in self.tape.items()},
            'head': self.head,
            'current_state': self.current_state,
            'halted': self.halted,
            'result': self.result
        }

    @classmethod
    def from_dict(cls, data):
        """Cria instância a partir de dicionário"""
        transitions = {}
        for k, v in data['transitions'].items():
            state, sym = k.split(',', 1)
            transitions[(state, sym)] = tuple(v)

        return cls(
            states=set(data['states']),
            input_symbols=set(data['input_symbols']),
            tape_symbols=set(data['tape_symbols']),
            blank=data['blank'],
            transitions=transitions,
            start_state=data['start_state'],
            accept_states=set(data['accept_states']),
            reject_states=set(data['reject_states']),
            tape={int(k): v for k, v in data['tape'].items()},
            head=data['head'],
            current_state=data['current_state'],
            halted=data['halted'],
            result=data['result']
        )


def parse_spec(spec_text: str):
    """Parser da DSL para Máquina de Turing"""
    try:
        lines = []
        for raw in spec_text.splitlines():
            s = raw.strip()
            if not s or s.startswith('#'):
                continue
            lines.append(s)
        text = '\n'.join(lines)

        if 'transitions:' not in text:
            return None, "Especificação precisa da seção 'transitions:'"
        head, body = text.split('transitions:', 1)

        header: Dict[str, str] = {}
        for line in head.splitlines():
            if ':' in line:
                k, v = [x.strip() for x in line.split(':', 1)]
                v_clean = v.strip()
                if (v_clean.startswith("'") and v_clean.endswith("'")) or (v_clean.startswith('"') and v_clean.endswith('"')):
                    v_clean = v_clean[1:-1]
                header[k.lower()] = v_clean

        required = ['states', 'blank', 'start', 'accept', 'reject']
        for r in required:
            if r not in header:
                return None, f"Campo obrigatório ausente: {r}"

        states = set([s.strip()
                     for s in header['states'].split(',') if s.strip()])
        blank = header['blank']
        if blank == "":
            blank = "_"
        if len(blank) != 1:
            return None, "O campo 'blank' deve conter exatamente 1 caractere (ex.: _ ou espaço)."

        start = header['start']
        accept = set([s.strip()
                     for s in header['accept'].split(',') if s.strip()])
        reject = set([s.strip()
                     for s in header['reject'].split(',') if s.strip()])

        transitions: Dict[Tuple[str, str], Transition] = {}
        tape_symbols: Set[str] = set([blank])
        input_symbols: Set[str] = set()

        for line in body.splitlines():
            if not line:
                continue
            if '->' not in line or ',' not in line:
                return None, f"Linha de transição inválida: {line}"
            left, right = [x.strip() for x in line.split('->', 1)]
            s_state, s_read = [x.strip() for x in left.split(',', 1)]
            n_state, s_write, s_move = [x.strip() for x in right.split(',', 2)]
            if s_move not in ('L', 'R', 'N'):
                return None, f"Movimento inválido em: {line}"
            transitions[(s_state, s_read)] = (n_state, s_write, s_move)
            tape_symbols.add(s_read)
            tape_symbols.add(s_write)
            if s_read != blank:
                input_symbols.add(s_read)

        tm = TuringMachine(
            states=states,
            input_symbols=input_symbols or set(['0', '1']),
            tape_symbols=tape_symbols,
            blank=blank,
            transitions=transitions,
            start_state=start,
            accept_states=accept,
            reject_states=reject,
        )
        return tm, None
    except Exception as e:
        return None, f"Erro ao parsear especificação: {e}"

File: core/turing_machine.js
class TuringMachine {
  constructor({
    states,
    inputSymbols,
    tapeSymbols,
    blank,
    transitions,
    startState,
    acceptStates,
    rejectStates,
    tape = {},
    head = 0,
    currentState = null,
    halted = false,
    result = null
  }) {
    this.states = new Set(states);
    this.inputSymbols = new Set(inputSymbols);
    this.tapeSymbols = new Set(tapeSymbols);
    this.blank = blank;
    this.transitions = transitions;
    this.startState = startState;
    this.acceptStates = new Set(acceptStates);
    this.rejectStates = new Set(rejectStates);
    this.tape = { ...tape };
    this.head = head;
    this.currentState = currentState;
    this.halted = halted;
    this.result = result;
  }

  reset(inputString) {
    this.tape = {};
    for (let i = 0; i < inputString.length; i++) {
      this.tape[i] = inputString[i];
    }
    this.head = 0;
    this.currentState = this.startState;
    this.halted = false;
    this.result = null;
  }

  read() {
    return this.tape[this.head] || this.blank;
  }

  write(symbol) {
    if (symbol === this.blank) {
      delete this.tape[this.head];
    } else {
      this.tape[this.head] = symbol;
    }
  }

  step() {
    if (this.halted) return;

    if (this.acceptStates.has(this.currentState)) {
      this.halted = true;
      this.result = 'ACCEPT';
      return;
    }

    if (this.rejectStates.has(this.currentState)) {
      this.halted = true;
      this.result = 'REJECT';
      return;
    }

    const sym = this.read();
    const key = `${this.currentState},${sym}`;

    if (!this.transitions[key]) {
      this.halted = true;
      this.result = 'NO_TRANSITION';
      return;
    }

    const [newState, writeSym, move] = this.transitions[key];
    this.write(writeSym);

    if (move === 'L') {
      this.head -= 1;
    } else if (move === 'R') {
      this.head += 1;
    } else if (move === 'N') {
      // Não move
    } else {
      throw new Error(`Movimento inválido: ${move}`);
    }

    this.currentState = newState;
  }

  run(maxSteps = 1000) {
    let steps = 0;
    while (!this.halted && steps < maxSteps) {
      this.step();
      steps++;
    }
    if (!this.halted && steps >= maxSteps) {
      this.halted = true;
      this.result = 'MAX_STEPS';
    }
    return steps;
  }

  windowCells(span = 25) {
    const left = this.head - span;
    const right = this.head + span;
    const cells = [];
    for (let i = left; i <= right; i++) {
      cells.push([i, this.tape[i] || this.blank, i === this.head]);
    }
    return cells;
  }

  toDict() {
    return {
      states: Array.from(this.states),
      input_symbols: Array.from(this.inputSymbols),
      tape_symbols: Array.from(this.tapeSymbols),
      blank: this.blank,
      transitions: this.transitions,
      start_state: this.startState,
      accept_states: Array.from(this.acceptStates),
      reject_states: Array.from(this.rejectStates),
      tape: this.tape,
      head: this.head,
      current_state: this.currentState,
      halted: this.halted,
      result: this.result
    };
  }

  static fromDict(data) {
    return new TuringMachine({
      states: data.states,
      inputSymbols: data.input_symbols,
      tapeSymbols: data.tape_symbols,
      blank: data.blank,
      transitions: data.transitions,
      startState: data.start_state,
      acceptStates: data.accept_states,
      rejectStates: data.reject_states,
      tape: typeof data.tape === 'object' ? data.tape : {},
      head: data.head,
      currentState: data.current_state,
      halted: data.halted,
      result: data.result
    });
  }
}

function parseSpec(specText) {
  try {
    const lines = specText
      .split('\n')
      .map(line => line.trim())
      .filter(line => line && !line.startsWith('#'));

    const text = lines.join('\n');

    if (!text.includes('transitions:')) {
      return { tm: null, error: "Especificação precisa da seção 'transitions:'" };
    }

    const [head, body] = text.split('transitions:');

    const header = {};
    for (const line of head.split('\n')) {
      if (line.includes(':')) {
        const [k, v] = line.split(':', 2).map(s => s.trim());
        let vClean = v.trim();
        if ((vClean.startsWith("'") && vClean.endsWith("'")) ||
          (vClean.startsWith('"') && vClean.endsWith('"'))) {
          vClean = vClean.slice(1, -1);
        }
        header[k.toLowerCase()] = vClean;
      }
    }

    const required = ['states', 'blank', 'start', 'accept', 'reject'];
    for (const r of required) {
      if (!header[r]) {
        return { tm: null, error: `Campo obrigatório ausente: ${r}` };
      }
    }

    const states = new Set(header.states.split(',').map(s => s.trim()).filter(s => s));
    let blank = header.blank;
    if (blank === '') blank = '_';
    if (blank.length !== 1) {
      return { tm: null, error: "O campo 'blank' deve conter exatamente 1 caractere (ex.: _ ou espaço)." };
    }

    const start = header.start;
    const accept = new Set(header.accept.split(',').map(s => s.trim()).filter(s => s));
    const reject = new Set(header.reject.split(',').map(s => s.trim()).filter(s => s));

    const transitions = {};
    const tapeSymbols = new Set([blank]);
    const inputSymbols = new Set();

    for (const line of body.split('\n')) {
      if (!line.trim()) continue;
      if (!line.includes('->') || !line.includes(',')) {
        return { tm: null, error: `Linha de transição inválida: ${line}` };
      }

      const [left, right] = line.split('->').map(s => s.trim());
      const [sState, sRead] = left.split(',', 2).map(s => s.trim());
      const parts = right.split(',').map(s => s.trim());

      if (parts.length !== 3) {
        return { tm: null, error: `Formato de transição inválido: ${line}` };
      }

      const [nState, sWrite, sMove] = parts;

      if (!['L', 'R', 'N'].includes(sMove)) {
        return { tm: null, error: `Movimento inválido em: ${line}` };
      }

      transitions[`${sState},${sRead}`] = [nState, sWrite, sMove];
      tapeSymbols.add(sRead);
      tapeSymbols.add(sWrite);
      if (sRead !== blank) {
        inputSymbols.add(sRead);
      }
    }

    const tm = new TuringMachine({
      states,
      inputSymbols: inputSymbols.size > 0 ? inputSymbols : new Set(['0', '1']),
      tapeSymbols,
      blank,
      transitions,
      startState: start,
      acceptStates: accept,
      rejectStates: reject
    });

    return { tm, error: null };
  } catch (e) {
    return { tm: null, error: `Erro ao parsear especificação: ${e.message}` };
  }
}

export { TuringMachine, parseSpec };
